### PR TITLE
[Design Review] Issue #1107: Serializer for clustered

### PR DESCRIPTION
**Pull request by ljacomet**
* Refactor `PersistableResourceService` and `LocalPersistenceService`
* Add `StateRepository`
* Update implicit `Serializer` constructor contract

I am interested in having comments on the high level design.

Future task in this context:
* Implement the `StateRepository` for disk and clustered
* Rework the `CompactPersistentJavaSerializer` and the `DefaultSerializationProvider`

### CODE REVIEW

File: api/src/main/java/org/ehcache/spi/persistence/StateRepository.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+   * <P>
+   *   If the map existed already, it is returned with its content fully available.
+   * </P>
+   *
+   * @param name the map name
+   * @param <K> the map key type, must be {@code Serializable}
+   * @param <V> the map value type, must be {@code Serializable}
+   * @return a map
+   */
+  <K, V> ConcurrentMap<K, V> getPersistentConcurrentMap(String name);
```
Comment (chrisdennis) [inline api/src/main/java/org/ehcache/spi/persistence/StateRepository.java]: Persistent is a property of the implementation, isn't it?

File: api/src/main/java/org/ehcache/spi/serialization/Serializer.java [hunk 2]
```diff
@@ -28,8 +28,9 @@
  * <dl>
  *   <dt><code><i>Serializer</i>(ClassLoader loader)</code>
  *   <dd>This constructor is used to initialize the serializer for transient caches.
```
Comment (chrisdennis) [inline api/src/main/java/org/ehcache/spi/serialization/Serializer.java]: This constructor is used to initialize 'stateless' serializers.

File: api/src/main/java/org/ehcache/spi/serialization/Serializer.java [hunk 3]
```diff
@@ -28,8 +28,9 @@
...
- *   <dt><code><i>Serializer</i>(ClassLoader loader, org.ehcache.core.spi.service.FileBasedPersistenceContext context)</code>
- *   <dd>This constructor is used to initialize the serializer for persistent caches.
+ *   <dt><code><i>Serializer</i>(ClassLoader loader, org.ehcache.spi.persistence.StateRepository stateRepository)</code>
+ *   <dd>This constructor is used to initialize the serializer for persistent caches and allows them to store any relevant
```
Comment (chrisdennis) [inline api/src/main/java/org/ehcache/spi/serialization/Serializer.java]: This constructor is used to initialize stateful serializer.

File: api/src/main/java/org/ehcache/spi/serialization/Serializer.java (hunk 2 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/spi/serialization/Serializer.java]: No, it is the constructor we use when no persistent tier is configured. Which is what makes this implicit constructor contract even more weird. A stateful serializer still needs to have this constructor if it is fit for transient caches.

File: api/src/main/java/org/ehcache/spi/serialization/Serializer.java (hunk 3 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/spi/serialization/Serializer.java]: No, it is the constructor we use when a persistent tier is configured. Which is what makes this implicit constructor contract even more weird. A statelessl serializer still needs to have this constructor if it is fit for persistent caches.

File: api/src/main/java/org/ehcache/spi/persistence/StateRepository.java (hunk 1 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/spi/persistence/StateRepository.java]: Well, it depends if we want to always provide a `StateRepository` or only in the context of a persistent cache.

**Comment by ljacomet** [comment]
Conversation with @lorban @chrisdennis @cljohnso @AbfrmBlr @albinsuresh concluded this is good wrt design - closing to work on proper implementation

**Pull request closed without merging**